File: packages/domain-ts/src/visibleRead.ts
/** UI samples, not downloads or ACKs, establish contiguous read eligibility. */
export interface ReadRow { sequence: number; own: boolean }
export class VisibleReadTracker {
  private since = new Map<number, number>()
  private qualified = new Set<number>()
  private sampledAt: number | undefined
  private position: number
  private rows: ReadRow[]
  constructor(readonly scope: string, baseline: number, rows: readonly ReadRow[], private readonly completeThrough: number) {
    if(!Number.isSafeInteger(baseline) || baseline<0 || !Number.isSafeInteger(completeThrough) || completeThrough<0) throw new Error('INVALID_READ_PROOF')
    this.position=baseline;this.rows=[...rows].sort((a,b)=>a.sequence-b.sequence)
    const seen=new Set<number>()
    for(const row of this.rows) {
      if(!Number.isSafeInteger(row.sequence) || row.sequence<=0 || seen.has(row.sequence)) throw new Error('INVALID_READ_PROOF')
      seen.add(row.sequence)
    }
  }
  cancel(): void {this.since.clear();this.sampledAt=undefined}
  sample(scope: string, now: number, visible: ReadonlySet<number>, foreground: boolean): number {
    if(scope!==this.scope || !foreground || !Number.isFinite(now)) {this.cancel();return this.position}
    if(this.sampledAt!==undefined && (now<this.sampledAt || now-this.sampledAt>200)) this.since.clear()
    this.sampledAt=now
    for(const sequence of this.since.keys()) if(!visible.has(sequence)) this.since.delete(sequence)
    for(const row of this.rows) if(visible.has(row.sequence)) {
      const start=this.since.get(row.sequence) ?? now;this.since.set(row.sequence,start)
      if(now-start>=300) this.qualified.add(row.sequence)
    }
    for(const row of this.rows) {
      if(row.sequence<=this.position) continue
      if(row.sequence>this.position+1 && row.sequence-1>this.completeThrough) break
      if(!row.own && !this.qualified.has(row.sequence)) break
      this.position=row.sequence
    }
    return this.position
  }
}

File: apps/flutter-prototype/lib/core/visible_read.dart
class ReadRow {
  const ReadRow(this.sequence, this.own);
  final int sequence;
  final bool own;
}

/// UI samples establish eligibility; transport events never call this core.
class VisibleReadTracker {
  VisibleReadTracker(
    this.scope,
    int baseline,
    List<ReadRow> rows,
    this.completeThrough,
  ) : _position = baseline,
      _rows = List.of(rows)..sort((a, b) => a.sequence.compareTo(b.sequence)) {
    if (baseline < 0 ||
        baseline > 9007199254740991 ||
        completeThrough < 0 ||
        completeThrough > 9007199254740991) {
      throw const FormatException('INVALID_READ_PROOF');
    }
    final seen = <int>{};
    for (final row in _rows) {
      if (row.sequence <= 0 ||
          row.sequence > 9007199254740991 ||
          !seen.add(row.sequence)) {
        throw const FormatException('INVALID_READ_PROOF');
      }
    }
  }
  final String scope;
  final int completeThrough;
  final List<ReadRow> _rows;
  int _position;
  final _since = <int, int>{};
  final _qualified = <int>{};
  int? _sampledAt;
  void cancel() {
    _since.clear();
    _sampledAt = null;
  }

  int sample(String source, int now, Set<int> visible, bool foreground) {
    if (source != scope || !foreground) {
      cancel();
      return _position;
    }
    if (_sampledAt != null && (now < _sampledAt! || now - _sampledAt! > 200)) {
      _since.clear();
    }
    _sampledAt = now;
    _since.removeWhere((sequence, _) => !visible.contains(sequence));
    for (final row in _rows) {
      if (visible.contains(row.sequence)) {
        final start = _since.putIfAbsent(row.sequence, () => now);
        if (now - start >= 300) _qualified.add(row.sequence);
      }
    }
    for (final row in _rows) {
      if (row.sequence <= _position) continue;
      if (row.sequence > _position + 1 && row.sequence - 1 > completeThrough) {
        break;
      }
      if (!row.own && !_qualified.contains(row.sequence)) break;
      _position = row.sequence;
    }
    return _position;
  }
}
